composer.json

```diff
@@ -77,5 +77,59 @@
             "ergebnis/composer-normalize": true
         },
         "sort-packages": true
+    },
+    "scripts": {
+        "cs:check": "php php-cs-fixer fix --dry-run",
+        "cs:diff": "git diff HEAD --name-only --diff-filter=ACMR -- '*.php' | xargs -r php php-cs-fixer fix --dry-run --diff",
+        "cs:diff-fix": "git diff HEAD --name-only --diff-filter=ACMR -- '*.php' | xargs -r php php-cs-fixer fix",
+        "cs:fix": "php php-cs-fixer fix",
+        "dev-tools:check": [
+            "./dev-tools/check_file_permissions.sh",
+            "./dev-tools/check_trailing_spaces.sh"
+        ],
+        "dev-tools:docs": "php dev-tools/doc.php",
+        "dev-tools:install": "composer -d dev-tools install",
+        "mess-detector": "php dev-tools/vendor/bin/phpmd . ansi phpmd.xml",
+        "phpstan": "php -d memory_limit=256M dev-tools/vendor/bin/phpstan analyse",
+        "qa": "@quality-assurance",
+        "quality-assurance": [
+            "Composer\\Config::disableProcessTimeout",
+            "@cs:check",
+            "@sa",
+            "@test"
+        ],
+        "sa": "@static-analysis",
+        "static-analysis": "@phpstan",
+        "test": "@test:all",
+        "test:all": [
+            "Composer\\Config::disableProcessTimeout",
+            "phpunit --testsuite all"
+        ],
+        "test:branch": "git diff master... --name-only --diff-filter=ACMR -- '*Test.php' | xargs -r phpunit",
```

Comment:
what's the use case?
like, I changed `tokens.php`. what will this command do?

Reply:
Nothing, `xargs` will get empty set and `-r` ensures that it won't fail on pipe

<img width="864" alt="image" src="https://user-images.githubusercontent.com/600668/224996475-06fe74c1-961f-418d-9b1e-fd9da13a3548.png">

This script is especially helpful because of time needed to run all tests. So you can easily run only tests for test files that were changed within your branch.

Reply:
i find the name of this command and your description what command does very confusing

Reply:
What about `test:modified`?

Reply:
@keradus what do you think about suggestion above? `test:modified` or `test:staged` (yeah, I believe you should know that there's a drawback in this approach because `ACMR` filter does not catch new files that are not staged, so maybe such name could be helpful).

Reply:
`test:changed`? I know it means the same as "modified", but would it be more common as it's used e.g. in 
![image](https://user-images.githubusercontent.com/9282069/236564051-551d911d-30df-4f6e-a230-9dc6a4e99700.png)

Reply:
Can we also test files not committed (from `git diff HEAD --name-only`)?

I mean: do you know is it possible to have `git diff [magic] | xargs -r phpunit` to grab all of the committed files, staged for commit and unstaged?

Reply:
`test:changed` is also possible 🙂. In terms of non-staged files, it's not possible with Git magic: `git diff` [does not handle untracked files](https://twitter.com/_Codito_/status/1611142380463210496). Files must be at least staged, they don't have to be committed. But in practice it's not a big problem, you most probably want to run this script before committing, so you can stage, test and commit.

Reply:
@keradus @kubawerlos my proposal:

- `test:staged` (instead of current `test:diff`) for running only tests before committing, "staged" refers to Git and explicitly states files must be staged in order to run them
- I would keep `test:branch` as is, because it refers to changes made in the branch. Any other `modified`, `changed`, `committed` wouldn't be descriptive enough, because all of them apply not only to branch, but also to staged area (`modified` and `changed`) or repo in general (`committed`).

PS. @staabm you said you like Composer scripts, so probably you would use it - which name is best from your perspective?

Reply:
usually I don't have such "diff oriented" test builds. sounds interessting though.
not sure we really need 2 separate commands. If possible I would only add a single one, to ease DX.

-----

alternativly it could make sense to look into parallel tests as it seems the suite is only run single threaded

Reply:
In terms of scripts, these 2 have different purpose: one is for usage before commit, second before push. These can be used in Git Hooks depending on habits.

Parallelisation is separate topic, just like linter optimisations - I want to make CI faster, but everything in its time 😉.

Reply:
I understand you have different use-cases but I wonder whether its worth having 2 scripts for them, or just have a single "generic" command which builds a current branch/PR

Reply:
Is it a problem to have choice, though? With atomic scripts it's possible to create local alias that would do `composer test:branch && composer test:staged`, it can be also defined as group script in Composer if we want it.

But maybe I misunderstood what you mean by "single 'generic' command which builds a current branch/PR" 😉.

Reply:
if there is only one command, its easier to find and as a user I don't need to think about why there are several and which one to choose

Reply:
Scripts' descriptions are added just for that IMHO - you run `composer` and you see what scripts are available and what they do. There's no magic, single script that could handle different purposes (run all tests, run only staged etc). These are shortcuts for commands that you would need to run anyway to achieve several scenarios 🙂.

Reply:
@keradus @kubawerlos can we settle something [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6839/files#discussion_r1186843022)?

Reply:
@keradus @kubawerlos Since ParaUnit was introduced (after rebase I've changed underlying commands) I believe we can drop `test:branch` and `test:diff` completely. With fast linter `composer test` takes about 30s, which is totally fine to run every time before commit/push. I'll remove this scripts right now.